#### Mayo 2024
[Canción mas escuchada en Colombia en el 2020 🎼🎼](hhttps://open.spotify.com/track/380HmhwTE2NJgawn1NwkXi?si=1020cd84a15b48dc)
# Entrega 3
## [Por: Juan Felipe Calderón Cárdenas]

###Sección 0

In [41]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
%load_ext sql

In [5]:
%sql  sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db

In [30]:
%%sql
SELECT * FROM song LIMIT 20

 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[('spotify:track:4SDywo3ykB2PEthjXvo1UP', 6, 0, 'Creedence Clearwater Revival', 'spotify:artist:3IYUhFvPQItj6xySrBmZkd', 'I Put A Spell On You', 'spotify:album:7oIftk0P8cViwNpNEdCYl2', 271786, 'Creedence Clearwater Revival'),
 ('spotify:track:229SqGbacrLSBfb9XfJczF', 6, 1, 'STRFKR', 'spotify:artist:2Tz1DTzVJ5Gyh8ZwVr6ekU', 'Bury Us Alive', 'spotify:album:5pfqk4oFycMVynOD8S2Cqs', 190360, 'Reptilians'),
 ('spotify:track:0orNm6qQT2B5WFi0l7l3PV', 6, 2, '311', 'spotify:artist:41Q0HrwWBtuUkJc7C1Rp6K', "Jackolantern's Weather", 'spotify:album:6VIeIM5tetuv7FrYxISffU', 204160, '311'),
 ('spotify:track:1hWQvA6oGVJ2mAVsZ59AaV', 6, 3, 'Nathaniel Rateliff & The Night Sweats', 'spotify:artist:02seUFsFQP7TH4hLrTj77o', 'S.O.B.', 'spotify:album:2UCyiayMoJOwBilPLQPQvK', 247612, 'Nathaniel Rateliff & The Night Sweats'),
 ('spotify:track:56SkdBKyR2zOkjk6wVFI9s', 6, 4, 'Toadies', 'spotify:artist:2QMMktUMtwMSwUZATX1cRO', 'Possum Kingdom', 'spotify:album:7tFTqVGHceSlXOaz8IylBN', 309573, 'Rubberneck'),
 ('spotify:track:6y20BV5L33R8YXM0YuI38N', 6, 5, 'Queens of the Stone Age', 'spotify:artist:4pejUc4iciQfgdX6OKulQn', 'No One Knows', 'spotify:album:4w3NeXtywU398NYW4903rY', 278746, 'Songs For The Deaf'),
 ('spotify:track:6xy6jNeNTYwjnKTDzMyHw2', 6, 6, 'The Cranberries', 'spotify:artist:7t0rwkOPGlDPEhaOcVtOt9', 'Zombie', 'spotify:album:2KsgTeLQXz7yDV1joGOd2L', 306440, 'No Need To Argue (The Complete Sessions 1994-1995)'),
 ('spotify:track:0rK3iUuoyQp6QReDPzwlYD', 6, 7, 'Misfits', 'spotify:artist:1cXi8ALPQCBHZbf0EgP4Ey', 'Monster Mash', 'spotify:album:6MzZSl19bVR3xWwFCOZyhg', 157000, 'Project 1950'),
 ('spotify:track:2k5lMt43DRIoEZPuswnULH', 6, 8, 'Misfits', 'spotify:artist:1cXi8ALPQCBHZbf0EgP4Ey', 'Ghouls Night Out - Live', 'spotify:album:4CWxS5ASY9eRE2zXP8BTRd', 102539, 'Evilive'),
 ('spotify:track:3gN4UFoOnRHlMl3K4kU5zG', 6, 9, 'Misfits', 'spotify:artist:1cXi8ALPQCBHZbf0EgP4Ey', 'Hybrid Moments', 'spotify:album:51tAz06EJxwhsk8uNfWxBo', 102266, 'Static Age'),
 ('spotify:track:37gYTPAIfSk7igC3cpf7BL', 6, 10, 'Willy Moon', 'spotify:artist:5aPH21KxzBBb7SZTpuITrC', 'Yeah Yeah', 'spotify:album:2S3gt2I3A1xWMdce85Bf8M', 164773, "Here's Willy Moon"),
 ('spotify:track:1xpd5XqLsrQDBa81ON2kOb', 6, 11, 'The Flys', 'spotify:artist:3brlG1zNhqIkYhdneCaPVi', 'Got You (Where I Want You)', 'spotify:album:6926E0NLdNZiEjK9ygGUCq', 242453, 'Holiday Man'),
 ('spotify:track:4g3Ax56IslQkI6XVfYKVc5', 6, 12, 'Tame Impala', 'spotify:artist:5INjqkS1o8h1imAzPqGZBb', 'The Less I Know The Better', 'spotify:album:0rxKf57PZvWEoU8v3m5W2q', 216318, 'Currents'),
 ('spotify:track:4VrWlk8IQxevMvERoX08iC', 6, 13, 'Sia', 'spotify:artist:5WUlDfRSoLAfcVSX1WnrxN', 'Chandelier', 'spotify:album:3xFSl9lIRaYXIYkIn3OIl9', 216120, '1000 Forms Of Fear'),
 ('spotify:track:6kwAbEjseqBob48jCus7Sz', 6, 14, 'Sia', 'spotify:artist:5WUlDfRSoLAfcVSX1WnrxN', 'Elastic Heart', 'spotify:album:3xFSl9lIRaYXIYkIn3OIl9', 257200, '1000 Forms Of Fear'),
 ('spotify:track:0hAMkY2kwdXPPDfQ1e3BmJ', 6, 15, 'Arctic Monkeys', 'spotify:artist:7Ln80lUS6He07XvHI8qqHH', 'Dancing Shoes', 'spotify:album:50Zz8CkIhATKUlQMbHO3k1', 141133, "Whatever People Say I Am, That's What I'm Not"),
 ('spotify:track:2c4QEzLpcIQ6vPRJ26XQqg', 8, 0, 'Catfish and the Bottlemen', 'spotify:artist:2xaAOVImG2O6lURwqperlD', 'Twice', 'spotify:album:6PrcsM0LNzEYEmxvLNTJQN', 196946, 'The Ride'),
 ('spotify:track:1MoDk4XDqdX25GXaRfXY2O', 8, 1, 'Catfish and the Bottlemen', 'spotify:artist:2xaAOVImG2O6lURwqperlD', '7', 'spotify:album:6PrcsM0LNzEYEmxvLNTJQN', 256306, 'The Ride'),
 ('spotify:track:5atlAt28h9xM5wxPspr5MU', 8, 2, 'Cage The Elephant', 'spotify:artist:26T3LtbuGT1Fu9m0eRq5X3', 'Whole Wide World - Unpeeled', 'spotify:album:4zpN41aGAYPWX3dzE19rH7', 205706, 'Unpeeled'),
 ('spotify:track:2XTkpF9T2PKvcLgamGJGx1', 8, 3, 'Red Hot Chili Peppers', 'spotify:artist:0L8ExT028jH3ddEcZwqJJ5', 'Goodbye Angels', 'spotify:album:43otFXrY0bgaq5fB3GrZj6', 268733, 'The Getaway')]

###Sección 1

Pregunta 5: Contar el  número total de canciones

In [18]:
%%sql
SELECT COUNT(track_name)
FROM song

 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[(529108,)]

Pregunta 6: Contar el número total de canciones únicas

In [20]:
%%sql
SELECT COUNT(DISTINCT track_name)
FROM song

 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[(114492,)]

Pregunta 9: Nombres de las 10 canciones mas largas en duración

In [25]:
%%sql
SELECT duration_ms
FROM song
ORDER BY duration_ms DESC
LIMIT 10

 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[(5279768,),
 (5062000,),
 (4556387,),
 (4497994,),
 (4471884,),
 (4436000,),
 (4316154,),
 (4276000,),
 (4195000,),
 (4120258,)]

In [29]:
%%sql
SELECT track_name, duration_ms
FROM song
WHERE duration_ms >=4120258
ORDER BY duration_ms DESC
LIMIT 10

 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[('Heavy Rains with Thunder and Lightning - Natural Sounds of Nature', 5279768),
 ('Faze DJ-Set 08 - Continuous DJ Mix', 5062000),
 ('Dancehall Mix Tape, Vol. 2 (Continuous Mix)', 4556387),
 ('Bonobo Late Night Tales Continuous Mix', 4497994),
 ('Club Life: Miami (Continuous DJ Mix)', 4471884),
 ('Blue Marlin Ibiza - Continuous DJ Mix 1', 4436000),
 ('Blue Marlin Ibiza - Continuous DJ Mix 2', 4316154),
 ('Mixtape 13', 4276000),
 ('4x4=12 - Continuous Mix', 4195000),
 ('Ocean Waves for Sleep', 4120258)]

###Sección 2:

Pregunta 1: Calcular la duración total de todas las canciones en una playlist específica

In [31]:
%%sql
SELECT COUNT(DISTINCT pid)
FROM song

 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[(8000,)]

In [39]:
%%sql
SELECT pid, COUNT(pid), SUM (duration_ms)
FROM song
GROUP BY pid
ORDER BY COUNT(pid) DESC
LIMIT 10

 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[(10211, 250, 56522074),
 (10160, 250, 65442850),
 (7063, 250, 56260788),
 (7278, 248, 59211102),
 (3149, 248, 57599896),
 (10063, 247, 58715653),
 (6627, 247, 61772580),
 (2500, 247, 56282170),
 (6892, 246, 58348450),
 (3626, 246, 52356692)]

In [38]:
%%sql
SELECT pid, COUNT(pid), SUM (duration_ms)
FROM song
WHERE pid=6


 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[(6, 16, 3408479)]

**Respuesta:** la duracion total de la playlist 6 es 3408479 ms

Pregunta 2: Obtener la duración promedio de las canciones de las 10 playlists con más canciones

In [47]:
%%sql
SELECT pid, COUNT(track_uri), AVG (duration_ms) AS Average
FROM song
GROUP BY pid
ORDER BY COUNT (track_uri) DESC
LIMIT 10

 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[(10211, 250, 226088.296),
 (10160, 250, 261771.4),
 (7063, 250, 225043.152),
 (7278, 248, 238754.4435483871),
 (3149, 248, 232257.64516129033),
 (10063, 247, 237715.1943319838),
 (6627, 247, 250091.4170040486),
 (2500, 247, 227863.03643724698),
 (6892, 246, 237188.82113821138),
 (3626, 246, 212832.081300813)]

Pregunta 5: Encontrar la duración más larga y más corta de todas las canciones

In [6]:
%%sql
SELECT track_name, MAX (duration_ms)
FROM song
GROUP BY track_name
ORDER BY MAX (duration_ms) DESC
LIMIT 10

 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[('Heavy Rains with Thunder and Lightning - Natural Sounds of Nature', 5279768),
 ('Faze DJ-Set 08 - Continuous DJ Mix', 5062000),
 ('Dancehall Mix Tape, Vol. 2 (Continuous Mix)', 4556387),
 ('Bonobo Late Night Tales Continuous Mix', 4497994),
 ('Club Life: Miami (Continuous DJ Mix)', 4471884),
 ('Blue Marlin Ibiza - Continuous DJ Mix 1', 4436000),
 ('Blue Marlin Ibiza - Continuous DJ Mix 2', 4316154),
 ('Mixtape 13', 4276000),
 ('4x4=12 - Continuous Mix', 4195000),
 ('Ocean Waves for Sleep', 4120258)]

In [9]:
%%sql
SELECT track_name, MAX (duration_ms)
FROM song
WHERE track_name= "Heavy Rains with Thunder and Lightning - Natural Sounds of Nature"

 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[('Heavy Rains with Thunder and Lightning - Natural Sounds of Nature', 5279768)]

In [11]:
%%sql
SELECT track_name, MIN (duration_ms)
FROM song
GROUP BY track_name
ORDER BY MIN (duration_ms) ASC
LIMIT 10

 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[('Another State', 0),
 ('Paris', 0),
 ('The Famm', 0),
 ('Oxygen - Radio Edit', 186),
 ('Secret Mission - From "Galavant"', 188),
 ('Gravity', 198),
 ('In Cold Blood', 206),
 ('Carousel', 231),
 ('Deadcrush', 231),
 ('Contradiction', 240)]

In [12]:
%%sql
SELECT track_name, MIN (duration_ms)
FROM song
WHERE track_name= "Another State"

 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[('Another State', 0)]

Pregunta 6: Calcular la suma total de duraciones de todas las canciones de un artista específico

In [13]:
%%sql
SELECT artist_name, SUM (duration_ms)
FROM song
WHERE artist_name= "Queens of the Stone Age"

 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[('Queens of the Stone Age', 21649771)]

###Sección 3:

Pregunta 1: Calcular la duración total y promedio de todas las canciones en una playlist específica

In [15]:
%%sql
SELECT pid, SUM (duration_ms), AVG (duration_ms)
FROM song
WHERE pid= 8

 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[(8, 9948921, 216280.89130434784)]

In [26]:
%%sql
SELECT pid, track_name, duration_ms, COUNT(track_name)
FROM song
where pid=8
GROUP BY pid, track_name
order by pid ASC


 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[(8, '7', 256306, 1),
 (8, 'A-Punk', 137760, 1),
 (8, 'All My Loving - Remastered', 127853, 1),
 (8, 'Cabron', 218373, 1),
 (8, 'Come Together - Remastered', 259946, 1),
 (8, 'Daydream Believer', 179613, 1),
 (8, 'Dog Days Are Over', 251840, 1),
 (8, 'Feel It Still', 163253, 1),
 (8, 'First', 200360, 1),
 (8, 'Goodbye Angels', 268733, 1),
 (8, 'Hard To Concentrate', 241840, 1),
 (8, 'Help! - Remastered', 139560, 1),
 (8, 'Hey Jude - Remastered 2015', 425653, 1),
 (8, 'I Could Die For You', 192826, 1),
 (8, 'I Want To Hold Your Hand - Remastered 2015', 145746, 1),
 (8, 'Kathleen', 160886, 1),
 (8, 'Lay Me Down', 217520, 1),
 (8, 'Left Hand Free', 173630, 1),
 (8, 'Let It Be - Remastered', 243026, 1),
 (8, 'Love Is Mystical', 214360, 1),
 (8, 'Make You Feel Better', 231733, 1),
 (8, 'Minor Thing', 217200, 1),
 (8, 'Mr. Blue Sky', 303373, 1),
 (8, 'On Melancholy Hill', 233866, 1),
 (8, 'Ordinary World', 182893, 1),
 (8, 'Outlaws', 302333, 1),
 (8, 'Rainbow', 218640, 1),
 (8, 'Red Red Wine - Edit', 183853, 1),
 (8, 'Reverend', 234080, 1),
 (8, 'Revolution - Remastered', 205293, 1),
 (8, 'She Moves In Her Own Way', 169306, 1),
 (8, 'Ship To Wreck', 234526, 1),
 (8, 'Soul Meets Body', 229413, 1),
 (8, 'Soul To Squeeze', 289533, 1),
 (8, 'Soundcheck', 261560, 1),
 (8, 'Take It or Leave It', 207320, 1),
 (8, 'The Night We Met', 208211, 1),
 (8, 'Tongue Tied', 218013, 1),
 (8, 'Too Much To Think', 234451, 1),
 (8, 'Twice', 196946, 1),
 (8, 'Waste A Moment', 183093, 1),
 (8, 'Welcome To Your Life', 227800, 1),
 (8, 'What Kind Of Man', 216176, 1),
 (8, 'Whole Wide World - Unpeeled', 205706, 1),
 (8, 'Yesterday - Remastered', 125666, 1),
 (8, 'Young Blood', 208853, 1)]

Pregunta 3: Contar la cantidad de playlists que contienen más de un álbum diferente

In [53]:
%%sql
WITH ALBUM_DIF AS (
    SELECT pid,
           COUNT(DISTINCT album_uri) AS album_count,
           CASE
               WHEN COUNT(DISTINCT album_uri) > 1 THEN 'MAYOR QUE UNO'
               ELSE 'MENOR QUE UNO'
           END AS albums
    FROM song
    GROUP BY pid
)
SELECT COUNT(*) AS total_records
FROM ALBUM_DIF;



 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[(8000,)]

Pregunta 6: Contar la cantidad total de playlists que contienen más de 10 canciones -> Utilizar un COUNT(*)

In [68]:
%%sql
WITH TRACK_DIF AS (
    SELECT pid,
           COUNT( track_uri ) AS trackNUM,
           CASE
               WHEN COUNT(track_uri) > 10 THEN 'MAYOR QUE DIEZ'
               ELSE 'MENOR O IGUAL QUE DIEZ'
           END AS counts
    FROM song
    GROUP BY pid
)
SELECT COUNT(*) AS total_tracks
FROM TRACK_DIF;


 * sqlite:////content/drive/MyDrive/Colab_Notebooks/music_smaller_10000.db
Done.


[(8000,)]